### PR TITLE
make descriptions uniform + add exceptions

### LINKED ISSUE #2409: timeout or no timeout?
@mbrenn commented on [Fri May 03 2019](https://github.com/dotnet/dotnet-api-docs.de-de/issues/25)

Invoke(Action, DispatcherPriority, CancellationToken, TimeSpan) defines timeout variable as 'minimum time before start' while other overloads define timeout variable 'as maximum time until end'. 
Seems to be wrong or confusing.

---
#### Dokumentdetails

⚠ *Bearbeiten Sie diesen Abschnitt nicht. Er ist für die Verknüpfung von docs.microsoft.com zum GitHub-Artikel erforderlich.*

* ID: 8adbf2d5-66d0-15ab-743a-21ec03b05295
* Version Independent ID: 5269116e-7645-b2ba-839a-1aca693431f8
* Content: [Dispatcher.Invoke Method (System.Windows.Threading)](https://docs.microsoft.com/de-de/dotnet/api/system.windows.threading.dispatcher.invoke?view=netframework-4.8#feedback)
* Content Source: [xml/System.Windows.Threading/Dispatcher.xml](https://github.com/dotnet/dotnet-api-docs.de-de/blob/live/xml/System.Windows.Threading/Dispatcher.xml)
* Product: **dotnet-api**
* Technology: **system.windows**
* GitHub Login: @dotnet-bot
* Microsoft Alias: **dotnetcontent**

### PR DESCRIPTION
**Pull request by mairaw**
Fixes #2409  

@OliaG @sambent who can review this change for WPF docs?

### CODE REVIEW

File: xml/System.Windows.Threading/Dispatcher.xml [hunk 1]
```diff
@@ -1077,6 +1077,10 @@
   
  ]]></format>
         </remarks>
+        <exception cref="T:System.ArgumentNullException">
+          <paramref name="method" /> is <see langword="null" />.</exception>
+        <exception cref="T:System.ArgumentOutOfRangeException">
+          <paramref name="timeout" /> is a negative number other than -1 milliseconds, which represents infinite wait, and you're invoking across threads.</exception>
```
Comment (rpetrusha) [inline xml/System.Windows.Threading/Dispatcher.xml]: This is the wrong place to add "which represents infinite wait". An infinite wait should be documented in the params section, not the exceptions. The params section should also note how negative numbers are interpreted in cross-thread and same-thread calls.

File: xml/System.Windows.Threading/Dispatcher.xml [hunk 2]
```diff
@@ -1063,8 +1063,8 @@
       </Parameters>
       <Docs>
         <param name="method">A delegate to a method that takes parameters specified in <paramref name="args" />, which is pushed onto the <see cref="T:System.Windows.Threading.Dispatcher" /> event queue.</param>
-        <param name="timeout">The maximum amount of time to wait for the operation to complete.</param>
-        <param name="args">An array of objects to pass as arguments to the given method. Can be <see langword="null" />.</param>
+        <param name="timeout">The minimum amount of time to wait for the operation to start. Once the operation has started, it will complete before this method returns.</param>
```
Comment (rpetrusha) [inline xml/System.Windows.Threading/Dispatcher.xml]: - I don't think that this definition is accurate -- it suggests a minimum amount of time must elapse before the operation can start. The timeout typically defines the amount of time that can elapse before the operation is aborted if it has not started.   
- The sentence "Once the operation has started..." doesn't seem relevant to the timeout parameter.

**Comment by carlossanlop** [comment]
@vatsan-madhavan @amit-kabra Maira pointed out these are WPF owned changes. Can you please review them?

File: xml/System.Windows.Threading/Dispatcher.xml (hunk 2 above)
Comment (SamBent) [reply xml/System.Windows.Threading/Dispatcher.xml]: - Ron is right.  Both the old wording and the new are wrong (and so are the comments in the code).  The correct wording is "The maximum amount of time to wait for the operation to start". 
- The point of the second sentence (lifted verbatim from code comment) is to clarify that the timeout's job is to abandon operations that spend too long waiting in the queue, not to interrupt operations that take too long to run.  That's worth mentioning.
- Same remarks apply to all the methods being changed here.

File: xml/System.Windows.Threading/Dispatcher.xml (hunk 1 above)
Comment (SamBent) [reply xml/System.Windows.Threading/Dispatcher.xml]: I agree.  
What really happens is that -1 represents infinite wait.  In a same-thread call, any other negative timeout is converted to -1, resulting in infinite wait.  In a cross-thread call, any other negative timeout throws ArgumentOutOfRangeException.  (This is for compat with 4.0, which behaved that way by accident of implementation.)

File: xml/System.Windows.Threading/Dispatcher.xml [hunk 3]
```diff
@@ -1660,6 +1688,12 @@
   
  ]]></format>
         </remarks>
+        <exception cref="T:System.ArgumentNullException">
+          <paramref name="callback" /> is <see langword="null" />.</exception>
+        <exception cref="T:System.ArgumentOutOfRangeException">
+          <paramref name="timeout" /> is a negative number other than -1 milliseconds, which represents an infinite wait.</exception>
```
Comment (SamBent) [inline xml/System.Windows.Threading/Dispatcher.xml]: This method also needs the verbiage about negative timeouts throwing in cross-thread calls but not in same-thread calls.

**Comment by gewarren** [comment]
@mairaw Do you want me to take this PR over and get it merged?

**Comment by mairaw** [comment]
Yes, that would be great @gewarren

**Comment by gewarren** [comment]
@BillWagner Is it okay if I take this one?

**Comment by gewarren** [comment]
[Preview link](https://review.docs.microsoft.com/en-us/dotnet/api/system.windows.threading.dispatcher.invoke?view=netcore-3.1&branch=pr-en-us-2442#System_Windows_Threading_Dispatcher_Invoke_System_Action_System_Windows_Threading_DispatcherPriority_System_Threading_CancellationToken_System_TimeSpan_).

File: xml/System.Windows.Threading/Dispatcher.xml [hunk 4]
```diff
@@ -1577,15 +1577,7 @@
...
-        <returns>To be added.</returns>
-        <remarks>
-          <format type="text/markdown"><![CDATA[  
-  
-## Remarks  
- The value returned by `callback`.  
-  
- ]]></format>
-        </remarks>
+        <returns>The value returned by <paramref name="callback" />.</returns>
```
Comment (gewarren) [inline xml/System.Windows.Threading/Dispatcher.xml:1580]: @sambent Can you verify that this return value is correct? It looks like it was incorrectly placed in the Remarks before.
